feat: Menu refactor to Fiori3

kavya-b opened the pull request:
## Related Issue
Closes SAP/fundamental-styles#571

## Description
Update Menu styles, documentation and tests to align with Fiori3 specs

- removed `fd-menu__group` classes, their playground tests and documentation examples related to group headers.
- added new classes for compact and mobile modes: `fd-menu--compact` and `fd-menu--mobile`. Default will be cozy tablet mode.
- moved `fd-menu__item--separated` from `<ul>` level to its own element `fd-menu__separator` in a `<span>` since separators are now used instead of group headers to show grouping.
- using existing class `fd-menu__title` for the text that will be used in a `<span>`
- added new class for  secondary icon `fd-menu__addon-after` that will be used similar to `fd-menu__addon-before`
- added a modifier class for the affordance arrow when used in a submenu `fd-menu__addon-after--submenu` to get the right(or left in RTL) navigation arrow in a smaller icon size. 
- added new class for submenu `fd-menu__sublist` at the `<ul>` level corresponding to `fd-menu__list` of the parent menu.
- added new class `fd-menu__link` that wraps the item name and associated icons, if any.
- added new class `has-child` that can be applied on the parent `__link` class to apply styles for the parent element that has a submenu.
- moved `fd-menu__item` class to the parent `<li>` tag.

## Breaking changes
- Removed `fd-menu__item--separated`. Instead use `fd-menu__separator` in its own `<span>` after the `<li>` item where you want the separator.
- Removed `fd-menu__group`. Separator itself acts a visual cue for grouping.
- Removed `fd-menu--addon-before`(not `fd-menu__addon-before`) as it is unnecessary now and will be covered by existing classes.
- Moved `fd-menu__item` class to the parent `<li>` tag. Use `fd-menu__link` class on the `<a>` tag instead.

For new classes added, please refer to the above section.

## Screenshots
> **NOTE:** If you've made any style changes, please provide appropriate screenshots (before and after) to help reviewers.
>
> To see examples of which screenshots to include, go to [Screenshot Examples](https://github.com/SAP/fundamental-styles/wiki/Pull-Request-Screenshot-Examples).
### Before:
![image](https://user-images.githubusercontent.com/53509521/75222665-bb8fdf80-57ca-11ea-9b7c-671e18d09570.png)
![image](https://user-images.githubusercontent.com/53509521/75222702-cf3b4600-57ca-11ea-9684-6ff9afb9a0b0.png)

### After:
![image](https://user-images.githubusercontent.com/53509521/75222710-d2cecd00-57ca-11ea-9048-04d58dc8c22d.png)
![image](https://user-images.githubusercontent.com/53509521/75222806-04479880-57cb-11ea-94e9-f39edb343cfb.png)

### Updated screenshots:
![image](https://user-images.githubusercontent.com/53509521/76381586-d8f0ac00-637b-11ea-94ae-26e20327e337.png)

![image](https://user-images.githubusercontent.com/53509521/76381552-c7a79f80-637b-11ea-9433-e459961954ea.png)

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SAP/fundamental-styles?pullRequest=702) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/SAP/fundamental-styles?pullRequest=702) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SAP/fundamental-styles?pullRequest=702) it.</sub>

JKMarkowski commented on src/menu.scss:
```diff
@@ -5,165 +5,303 @@
 
-    text-decoration: none;
+    @include fd-hover() {
+      .#{$block}__option-name {
```
Could you change this element name to `__title`?

kavya-b replied:
done

JKMarkowski commented on src/menu.scss:
```diff
@@ -5,165 +5,303 @@
     list-style: none;
-    margin-bottom: 0;
-    padding-left: 0;
+    max-width: 20rem;
```
Let's put `max-width` to parent `fd-menu`  container

kavya-b replied:
have moved to fd-menu

JKMarkowski commented on src/menu.scss:
```diff
@@ -5,165 +5,303 @@
+
+      @include fd-rtl() {
+        margin: 0 -0.75rem 0 0; // since before and after icons switch places, their margins also switch
+      }
```
Let's not use `fd-screen`, we let users to add size modes by themselves on `fd-menu` level. so ex. `fd-menu--compact`.

kavya-b replied:
I was handling cozy, compact modes for tablet, desktop and mobile through this. If we provide option to user to add size modes, then we won't have control if they decide to use compact mode on phone thereby not aligning to Fiori3, right?

JKMarkowski replied:
Actually, in all of cases, when we have more than 1 `size mode` such as cozy/compact, we let users decide on application level, which one they want to use it.

JKMarkowski replied:
Actually, in all of cases, when we have more than 1 `size mode` such as cozy/compact, we let users decide on application level, which one they want to use it.

kavya-b replied:
added `fd-menu--compact-desktop` and `fd-menu--mobile` classes. Tablet cozy is default.

JKMarkowski commented on src/menu.scss:
```diff
@@ -5,165 +5,303 @@
+    flex: 0 0 auto;
+    display: flex;
     background-color: transparent;
     float: left;
```
Try to remove floats, it's out of control. There are better alternatives

kavya-b replied:
removed

JKMarkowski commented on src/menu.scss:
```diff
@@ -5,165 +5,303 @@
+    }
+
+    background-color: transparent;
+    float: right;
```
Again float, which is not needed

kavya-b replied:
removed

JKMarkowski commented on src/menu.scss:
```diff
@@ -5,165 +5,303 @@
+      }
+    }
+
+    flex: 0 0 auto;
```
It's also not needed I think.

kavya-b replied:
removed

kavya-b commented:
> ### The menu item has text decorator this doesn't match fiori 3
> ![image](https://user-images.githubusercontent.com/10849982/75440951-0c374200-595d-11ea-8c4b-1d789255a239.png)

**This is coming from the styles to <a> tag in documentation itself. It appears fine in playground:**
![image](https://user-images.githubusercontent.com/53509521/75446257-11ac8080-598d-11ea-95de-340da6e412a3.png)
 
> ### wrong width in IE 11
> ![image](https://user-images.githubusercontent.com/10849982/75442462-d34c9c80-595f-11ea-9d39-427c9ddb6598.png)

I will work on this; however, in the meanwhile, do you think this is related to https://github.com/SAP/fundamental-styles/pull/691 ?

> ### Could you add RTL example to documentation?

I will add this.

kavya-b commented:
> * Let's add cozy mode by default, because right now there is only `compact` mode.

Do you mean, I should have the cozy mode applied by default? But for desktop, it is mentioned as Compact mode, right?
> * There are also differencted between Compact/Cozy(tablet)/Cozy(Phone)
>   ![image](https://user-images.githubusercontent.com/26483208/75438374-1efb4800-5958-11ea-9847-b7eb02ad6c23.png)

Currently, the styles of cozy, compact are applied when the screen is resized. However, from your other comment, it seems that these styles are better provided as separate `fd-menu-cozy` `fd-menu-compact` classes. I will work on changing this.
> * Background color of `fd-menu` is not set

I assumed that since fd-menu will be part of a popover, the background color, box shadow and border attributes of the container should be part of the popover. I have verified that the appropriate values are applied to the popover class. 
> * Remove underline from `<a>` if used with `__title`
>   ![image](https://user-images.githubusercontent.com/26483208/75438538-68e42e00-5958-11ea-93d9-08c2eab29732.png)

This is coming from the styles `<a>` tag in documentation itself. It appears fine in playground:
![image](https://user-images.githubusercontent.com/53509521/75447165-d27f2f00-598e-11ea-98b5-94c42c629d95.png)
Currently same thing is happening for side-navigation component as well.

rengare commented:
> I will work on this; however, in the meanwhile, do you think this is related to #691 ?
Frankly, I don't know, we know that it is due to the popover and input used together but maybe it is related. I would have to check that

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/SAP/fundamental-styles?pullRequest=702) <br/>All committers have signed the CLA.

kavya-b commented on src/menu.scss:
```diff
@@ -5,165 +5,303 @@
+    .#{$block}__option-name {
+      @include fd-reset();
+
+      flex: 3 3 10%;
```
we need this to ensure the primary, secondary icons and title inside a menu item are properly placed. If removed, the secondary icon comes to the end of the title and misaligns.

kavya-b commented:
There are 2 known issues in IE as of now that I couldn’t find a fix for but I also see the same issue is present in components that use List component(Select, Multi Input etc)-
1. focus border outline-offset property is not applied properly in IE due to which it sometimes does not appear properly
2. clicking on the parent <li>  area alone activates the states (active, selected etc). If you click on an area that is a child of <li> the states don’t show up.

stefanoScalzo commented:
![Screen Shot 2020-03-04 at 8 05 15 PM](https://user-images.githubusercontent.com/50607147/75937640-83843e80-5e53-11ea-8451-ab619991c19a.png)
disabled should not change color on hover and on click

kavya-b commented:
> ![Screen Shot 2020-03-04 at 8 05 15 PM](https://user-images.githubusercontent.com/50607147/75937640-83843e80-5e53-11ea-8451-ab619991c19a.png)
> disabled should not change color on hover and on click

fixed

JKMarkowski commented on src/menu.scss:
```diff
@@ -5,165 +5,295 @@
-    display: block;
+    display: flex;
+    align-items: center;
     transition: all $fd-animation-speed ease-in;
```
There is no mention about transition in menu specs

kavya-b replied:
Missed this. Will remove it, thanks for pointing it out!

InnaAtanasova commented on src/menu.scss:
```diff
@@ -5,165 +5,295 @@
+    background-color: transparent;
+
+    @include fd-icon-size("m", "before");
+    @include fd-icon-size("m", "after");
```
This mixin makes the icon 16px. We should use `rem`
<img width="496" alt="Screen Shot 2020-03-09 at 1 50 57 PM" src="https://user-images.githubusercontent.com/39598672/76242433-200c5f00-620d-11ea-98e5-33cb418f9907.png">
<img width="264" alt="Screen Shot 2020-03-09 at 1 50 49 PM" src="https://user-images.githubusercontent.com/39598672/76242439-200c5f00-620d-11ea-84c4-03c5e02ca021.png">

kavya-b replied:
Okay, I will just use `var(--sapFontLargeSize)` directly instead

InnaAtanasova commented on src/menu.scss:
```diff
@@ -5,165 +5,295 @@
-    width: 30px;
+
+    @include fd-icon-size("s", "before");
+    @include fd-icon-size("s", "after");
```
This size should be applied only if the user wants affordance arrow. It shouldn't be applied for all addon-after. Maybe create a modifier class to handle the affordance arrow case

kavya-b replied:
so, would you say `fd-menu__addon-after` by default takes any icon in the normal font-size of 1rem but for affordance arrow I should keep a class like `fd-menu__addon-after--submenu` and handle the smaller size case here?

kavya-b commented:
> <img alt="Screen Shot 2020-03-09 at 2 01 41 PM" width="649" src="https://user-images.githubusercontent.com/39598672/76243327-8cd42900-620e-11ea-96c6-9af485308036.png">
> 
> I don't think it's a good approach to add the separator to the item. It should be a separate element. In the case where we have an item that is a separator and is selected, the selected border overwrites the separator. It should be like this: (has both lines)
> <img alt="Screen Shot 2020-03-09 at 2 01 28 PM" width="1297" src="https://user-images.githubusercontent.com/39598672/76243555-e3d9fe00-620e-11ea-9395-bebb342644d9.png">

I see. I will add a separate element independent of the item that is placed after the item.
> 
> * `fd-menu--compact-desktop` should be `fd-menu--compact` for consistency with the other components

Will change this
> * In Mobile mode how do we display the cascading (the second menu level) ?
> 
> <img alt="Screen Shot 2020-03-09 at 2 05 50 PM" width="1226" src="https://user-images.githubusercontent.com/39598672/76243704-1a177d80-620f-11ea-8a1c-87f94e1ffd5c.png">
> 
The spec says _'Both controls are cascading, except for the sap.m.Menu on phone, where it opens fullscreen (with drill-drown navigation).'_ So the submenu will look the same way as the parent menu. 
> * For the documentation part - would you please remove the `Affordance arrow` if there is no second level. It gives the impression that something is not working because clicking on it does not open anything.

I will remove the arrows and instead show other icons to show how secondary icons look.

kavya-b commented:
Addressed @JKMarkowski and @InnaAtanasova's comments: 
- removed redundant right-arrow icons when submenu was not really present. For submenu, use `fd-menu__addon-after--submenu` class modifier to get the smaller size for affordance arrow:
`<span class="fd-menu__addon-after fd-menu__addon-after--submenu sap-icon--navigation-right-arrow"></span>`
- instead used other icons to showcase secondary icon size
![image](https://user-images.githubusercontent.com/53509521/76381965-173a9b00-637d-11ea-95e1-7fb5c2f51afa.png)
- using separator in a different element
![image](https://user-images.githubusercontent.com/53509521/76382019-4ea94780-637d-11ea-8246-706a6f153a33.png)
- removed 'desktop' from modifier class name
- removed transition property
- applied font-size on icon directly
- updated playground tests and screenshots accordingly - added selected-separated usecase to visual test and documentation
![image](https://user-images.githubusercontent.com/53509521/76382300-44d41400-637e-11ea-9c5e-3d06ee033643.png)

InnaAtanasova commented:
<img width="633" alt="Screen Shot 2020-03-11 at 11 08 11 AM" src="https://user-images.githubusercontent.com/39598672/76431903-9e414080-6388-11ea-87d7-3e3d1ba795a5.png">

Hi @kavya-b , 
`fd-menu__item--separated` is contradicting the BEM methodology. The separator is not an menu item, which means we can not use `fd-menu__item`. Since it's not a menu item and is not a modifier of this item we can not use the `--separated`. You can rename it to something like `fd-menu__separator`.
For more info about BEM you can check [this](http://getbem.com/introduction/)

kavya-b commented:
> <img alt="Screen Shot 2020-03-11 at 11 08 11 AM" width="633" src="https://user-images.githubusercontent.com/39598672/76431903-9e414080-6388-11ea-87d7-3e3d1ba795a5.png">
> 
> Hi @kavya-b ,
> `fd-menu__item--separated` is contradicting the BEM methodology. The separator is not a menu item, which means we can not use `fd-menu__item`. Since it's not a menu item and is not a modifier of this item we can not use the `--separated`. You can rename it to something like `fd-menu__separator`.
> For more info about BEM you can check [this](http://getbem.com/introduction/)

Okay, have changed the name to `fd-menu__separator` now.

droshev commented:
@kavya-b Few comments:
- can you update the documentation and create examples without edge cases like long text items; hybrid items of text and icons - those can be added as part of the playground so can have testing examples there. The goal of the documentation is to show correct and recommended layouts without too long text or inconsistent icon/texts
- would you mind documenting the class names that the menu supports (e.g. `fd-menu__addon-before`, `fd-menu__addon-after`, `fd-menu__separator`,  all state classes, `fd-menu__sublist`)

droshev commented on src/menu.scss:
```diff
@@ -5,165 +5,294 @@
+      }
+
+      .#{$block}__addon-after {
+        // border-bottom-color: var(--sapList_BorderColor); // TODO change this when fade() evaluates to a value
```
you can remove this comment

kavya-b replied:
done. have pushed in the other changes as well, please check.

kavya-b commented:
> @kavya-b Few comments:
> 
> * can you update the documentation and create examples without edge cases like long text items; hybrid items of text and icons - those can be added as part of the playground so can have testing examples there. The goal of the documentation is to show correct and recommended layouts without too long text or inconsistent icon/texts

okay, so for documentation, instead of showing every combination of single icon/double icon/long text, I keep only one format and use the same even for mobile mode and rtl mode? Should I keep the example with different states(active, hover, selected etc) or remove it as well?

> * would you mind documenting the class names that the menu supports (e.g. `fd-menu__addon-before`, `fd-menu__addon-after`, `fd-menu__separator`,  all state classes, `fd-menu__sublist`) - you can get inspired from here: https://fundamental-styles.netlify.com/components/bar.html

had added some of these in the PR description, but will update to the documentation description as well.
> * in the PR add a section `BREAKING CHANGES` and add all class names that were removed

Okay, will do this.

droshev commented:
@kavya-b thanks for the updates.
Could you try to review your component following the points from https://github.com/SAP/fundamental-styles/wiki/Self-Contained-Styles. There is something wrong with the separator:
![Screen Shot 2020-03-18 at 2 31 03 PM](https://user-images.githubusercontent.com/4380815/76994963-838d3f80-6925-11ea-98ba-99e969b95ebe.png)

kavya-b commented:
> https://github.com/SAP/fundamental-styles/wiki/Self-Contained-Styles

Hi @droshev, I tried to figure out if I am missing something that makes the unnormalized mode look like that, but cannot point to anything.
`fd-menu__separator` is in its own element in order to cover the edge case where we have a selected item and also need to show the separator, as pointed by @InnaAtanasova  in this comment: https://github.com/SAP/fundamental-styles/pull/702#pullrequestreview-371413725

droshev commented:
@kavya-b We need to check how the nested menu behaves and looks on mobile. I guess it may behave like the nested list in side navigation. Can you follow up on that?
Meanwhile i fixed the self-contained issue I had raised before.

kavya-b commented:
> @kavya-b We need to check how the nested menu behaves and looks on mobile. I guess it may behave like the nested list in side navigation. Can you follow up on that?

Hi @droshev , the [spec](https://wiki.wdf.sap.corp/wiki/pages/viewpage.action?pageId=2061501144#Menu(Fiori3)-MenuInteractionStates) says "_On Phone the Submenu opens in a new page, leaving the Menu option behind._ " The same visual design will apply for the submenu; on clicking of the parent item, the submenu will open in the same dialog, replacing the parent menu options, and the device's back button will bring it back to the parent menu. 
Does this need to be added to the documentation? I am not sure how to show a full screen replacement of parent menu in the same example using only aria-controls. I have something like this now: 
![image](https://user-images.githubusercontent.com/53509521/77290106-6da2c480-6d01-11ea-988e-f2907b11a222.png)
Please let me know if this suffices, I will push in the changes.
> Meanwhile i fixed the self-contained issue I had raised before.

Thank you for this.

kavya-b commented:
@droshev As discussed, the documentation for mobile mode has been modified to reflect full screen:
![submenu_mobile](https://user-images.githubusercontent.com/53509521/77606577-e3936f80-6f3d-11ea-9d53-d3847e207480.gif)
